### PR TITLE
Add option_and_then_some lint

### LINKED ISSUE #4299: New lint: and_then_some
Name: and_then_some
Group: complexity

Warn about `option.and_then(|o| Some(x))` and suggest replacing with `option.map(|o| x)`.

### PR DESCRIPTION
**Pull request by lzutao**
changelog: Add complexity lint to warn about `option.and_then(|o| Some(x))` and suggest replacing with `option.map(|o| x)`.

Closes #4299

### CODE REVIEW

**Comment by flip1995** [comment]
It seems that this lint would fit in the `methods` module.

**Comment by lzutao** [comment]
Looks like this lint conflicts with tests in 
https://github.com/rust-lang/rust-clippy/blob/34457fbd8b0980e0d41e34372cb99f86490d07f7/tests/ui/option_map_or_none.rs#L6-L13

Should I fix those test?

**Comment by flip1995** [comment]
2 step linting is totally fine. So I think adding an `allow(option_and_then_some)` to the test file is the best solution here.

**Comment by lzutao** [comment]
What about linting again `.map_or(None, |x| y)` and suggesting to use`map(|x| y)`
instead?

**Comment by flip1995** [comment]
That would also be possible. I don't really know how this lint is implemented, so this is a shot in the dark: My concern with this is, that the lint of `map_or(None, expr)` (porbably) just copies the `expr` of the closure and builds a suggestion from there. Directly suggesting `map(expr')` would require to change every return of `expr` of type `Option<T>` to a return of type `T`. Which obviously is way harder to do.

What could be done is, to add a special case to `option_map_or_none` that suggests `map` directly in the easy cases (similar to your implementation of this lint)

**Comment by lzutao** [comment]
Blocked on #4395

File: clippy_lints/src/methods/mod.rs [hunk 1]
```diff
@@ -2052,6 +2080,57 @@ fn lint_map_or_none<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx hir::Expr,
...
+            let closure_expr = remove_blocks(&closure_body.value);
+
+            // let note = format!("{:?}", args[1]);
+            // span_note_and_lint(cx, OPTION_AND_THEN_SOME, closure_args_span, "Outer debugging
+            // information", closure_args_span, &note);
+
+            if let hir::ExprKind::Call(ref some_expr, ref some_args) = closure_expr.node {
+                if let hir::ExprKind::Path(ref qpath) = some_expr.node {
+                    if match_qpath(qpath, &paths::OPTION_SOME) {
+                        if some_args.len() == 1 {
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs]: You can use an `if_chain!` here

File: tests/ui/option_and_then_some.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// run-rustfix
+#![deny(clippy::option_and_then_some)]
+
+// need a main anyway, use it get rid of unused warnings too
+pub fn main() {
+    let x = Some(5);
+    // the easiest cases
+    x.and_then(Some);
```
Comment (flip1995) [inline tests/ui/option_and_then_some.rs]: This is a noop, can we suggest `x` in this case?

Comment (lzutao) [reply tests/ui/option_and_then_some.rs]: Done.

**Comment by lzutao** [comment]
Ready for another round of review! :tada:

File: clippy_lints/src/methods/mod.rs [hunk 3]
```diff
@@ -2052,6 +2080,63 @@ fn lint_map_or_none<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx hir::Expr,
     }
 }
 
+/// Lint use of `_.and_then(|x| Some(y))` for `Option`s
+fn lint_option_and_then_some(cx: &LateContext<'_, '_>, expr: &hir::Expr, args: &[hir::Expr]) {
+    const LINT_MSG: &str = "using `Option.and_then(|x| Some(y))`, which is more succinctly expressed as `map(|x| y)`";
+    const NO_OP_MSG: &str = "using `Option.and_then(Some)`, which is no-op";
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs]: NIT:

```suggestion
    const NO_OP_MSG: &str = "using `Option.and_then(Some)`, which is a no-op";
```

File: tests/ui/option_and_then_some.rs [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    let x = Some(5);
+    // the easiest cases
+    let _ = x.and_then(Some);
+    let _ = x.and_then(|o| Some(o + 1));
+    // and an easy counter-example
+    let _ = x.and_then(|o| if o < 32 { Some(o) } else { None });
+
+    // Different type
+    let x: Result<u32, &str> = Ok(1);
+    let _ = x.and_then(Ok);
```
Comment (flip1995) [inline tests/ui/option_and_then_some.rs]: Couldn't be the lint be applied to `Result` as well?

**Comment by lzutao** [comment]
So I have an counter example which we do not want to lint:
```rust
fn foo() -> Option<String> {
    let x = Some(String::from("hello"));
    Some("hello".to_owned()).and_then(|s| Some(format!("{}{}", s, x?)))
}
```

How do I check if expression in `Some` have `try` block?

**Comment by flip1995** [comment]
You can write a visitor, that walks the expression inside the `Some(_)` and searches for return expressions.

You can search the Clippy repo on how to implement the Iterator or have a look at the documentation: https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/intravisit/trait.Visitor.html

**Comment by lzutao** [comment]
Tests passed locally.

File: tests/ui/option_and_then_some.rs (hunk 4 above)
Comment (lzutao) [reply tests/ui/option_and_then_some.rs]: I think we should open a different issue and PR.

**Comment by mikerite** [comment]
I made a Visitor like that in #3427. I prefer this one but you should probably check for `return` expressions as well. Does the lint handle this case correctly: 
```
pub fn example2(x: bool) -> Option<&'static str> {
    Some("a").and_then(|s| Some(if x { "b" } else { return None } ))
}

fn main() {
    println!("{:?}", example2(true));
    println!("{:?}", example2(false));
}```

We should probably make a common function for this in `util` but maybe not in this PR.

**Comment by flip1995** [comment]
Thanks!

@bors r+

**Pull request merged**